File: core/src/main/java/com/nagopy/android/overlayviewmanager/OverlayView.java
/*
 * Copyright 2017 75py
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.nagopy.android.overlayviewmanager;

import android.app.Activity;
import android.graphics.PixelFormat;
import android.os.Build;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;

import com.nagopy.android.overlayviewmanager.internal.Logger;
import com.nagopy.android.overlayviewmanager.internal.OverlayWindowManager;
import com.nagopy.android.overlayviewmanager.internal.ScreenMonitor;

import androidx.annotation.FloatRange;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;
import androidx.core.view.GravityCompat;
import androidx.core.view.ViewCompat;

import static android.view.ViewGroup.LayoutParams.WRAP_CONTENT;
import static android.view.WindowManager.LayoutParams.FLAG_LAYOUT_INSET_DECOR;
import static android.view.WindowManager.LayoutParams.FLAG_LAYOUT_IN_SCREEN;
import static android.view.WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS;
import static android.view.WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE;
import static android.view.WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE;
import static android.view.WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL;
import static android.view.WindowManager.LayoutParams.TYPE_APPLICATION;
import static android.view.WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY;
import static android.view.WindowManager.LayoutParams.TYPE_SYSTEM_ALERT;
import static android.view.WindowManager.LayoutParams.TYPE_SYSTEM_ERROR;
import static android.view.WindowManager.LayoutParams.TYPE_SYSTEM_OVERLAY;

/**
 * The wrapper class of the view to overlay.
 * This class allows a view to be displayed as an overlay on top of other content.
 *
 * @param <T> The type of the view to be overlaid.
 */
public class OverlayView<T extends View> {

    @VisibleForTesting
    T view;
    @VisibleForTesting
    WindowManager.LayoutParams params;
    @VisibleForTesting
    boolean isVisible;

    @VisibleForTesting
    boolean isDraggable;
    @VisibleForTesting
    boolean isTouchable;
    @VisibleForTesting
    boolean allowViewToExtendOutsideScreen;

    @VisibleForTesting
    int lastWindowType;

    @VisibleForTesting
    OverlayWindowManager overlayWindowManager;

    @VisibleForTesting
    ScreenMonitor screenMonitor;

    enum ViewScope {APPLICATION, ACTIVITY}

    @VisibleForTesting
    ViewScope viewScope;

    /**
     * Create a new instance with ViewScope set to APPLICATION.
     * This is used when the overlay is to be displayed regardless of the activity lifecycle.
     *
     * @param view The view to overlay.
     * @param <T>  The type of the view.
     * @return A new instance of OverlayView.
     */
    @NonNull
    static <T extends View> OverlayView<T> create(@NonNull T view) {
        OverlayView<T> overlayView = new OverlayView<>(view);
        overlayView.viewScope = ViewScope.APPLICATION;
        overlayView.overlayWindowManager = OverlayWindowManager.getApplicationInstance();
        return overlayView;
    }

    /**
     * Create a new instance with ViewScope set to ACTIVITY.
     * This is used when the overlay should be managed within an activity's lifecycle.
     *
     * @param view     The view to overlay.
     * @param <T>      The type of the view.
     * @param activity The parent activity.
     * @return A new instance of OverlayView.
     */
    @NonNull
    static <T extends View> OverlayView<T> create(@NonNull T view, Activity activity) {
        OverlayView<T> overlayView = new OverlayView<>(view);
        overlayView.viewScope = ViewScope.ACTIVITY;
        overlayView.overlayWindowManager = OverlayWindowManager.getActivityInstance(activity);
        return overlayView;
    }

    /**
     * Private constructor.
     * Use {@link OverlayView#create(View)} or {@link OverlayView#create(View, Activity)} to get an instance.
     *
     * @param view The view to overlay.
     */
    private OverlayView(@NonNull T view) {
        this.view = view;
        params = new WindowManager.LayoutParams();
        params.format = PixelFormat.TRANSLUCENT;
        setWidth(WRAP_CONTENT)
                .setHeight(WRAP_CONTENT)
                .setTouchable(false)
                .setDraggable(false)
                .allowViewToExtendOutsideScreen(false)
                .setGravity(Gravity.TOP | Gravity.START);
        isVisible = false;
        screenMonitor = ScreenMonitor.getInstance();
    }

    /**
     * Starts overlaying the view.
     * If the view is already shown, this method does nothing.
     *
     * @return This object for method chaining.
     */
    @NonNull
    public synchronized OverlayView<T> show() {
        if (!isVisible()) {
            updateParams();

            overlayWindowManager.show(view, params);
            isVisible = true;
            lastWindowType = params.type;

            if (isDraggable) {
                screenMonitor.request(this);
            }
        }
        return this;
    }

    /**
     * Requests updating the window parameters.
     * If the view is shown, it will call {@link WindowManager#updateViewLayout(View, ViewGroup.LayoutParams)}.
     */
    public synchronized void update() {
        if (isVisible()) {
            updateParams();

            if (lastWindowType == params.type) {
                overlayWindowManager.update(view, params);
            } else {
                Logger.d("Window type is changed. %d => %d", lastWindowType, params.type);
                overlayWindowManager.hide(view);
                overlayWindowManager.show(view, params);
                lastWindowType = params.type;
            }

            if (isDraggable) {
                screenMonitor.request(this);
            } else {
                screenMonitor.cancel(this);
            }
        }
    }

    /**
     * Stops overlaying the view.
     * If the view is already hidden, this method does nothing.
     */
    public synchronized void hide() {
        if (isVisible()) {
            overlayWindowManager.hide(view);
            isVisible = false;

            if (isDraggable) {
                screenMonitor.cancel(this);
            }
        }
    }

    /**
     * Returns the managed view.
     *
     * @return The overlaid view.
     */
    @NonNull
    public T getView() {
        return view;
    }

    /**
     * Returns the visibility status of the view.
     *
     * @return true if the view is visible, false otherwise.
     */
    public boolean isVisible() {
        return isVisible;
    }

    /**
     * Registers a callback to click events.
     *
     * @param onClickListener The callback to be invoked on click events.
     * @return This object for method chaining.
     */
    @NonNull
    public OverlayView<T> setOnClickListener(@Nullable View.OnClickListener onClickListener) {
        view.setOnClickListener(onClickListener);
        return this;
    }

    /**
     * Enables or disables touch events for this view.
     *
     * @param touchable true to enable touch events, false to disable.
     * @return This object for method chaining.
     */
    @NonNull
    public OverlayView<T> setTouchable(boolean touchable) {
        this.isTouchable = touchable;
        return this;
    }

    /**
     * Allows or disallows the view to extend outside the screen.
     * If true, the view can overlay on a navigation bar or a status bar.
     * Note: Using this option may cause unexpected behavior.
     *
     * @param allowViewToExtendOutsideScreen true to allow extending, false otherwise.
     * @return This object for method chaining.
     */
    @NonNull
    public OverlayView<T> allowViewToExtendOutsideScreen(boolean allowViewToExtendOutsideScreen) {
        this.allowViewToExtendOutsideScreen = allowViewToExtendOutsideScreen;
        return this;
    }

    /**
     * Enables or disables moving the view by dragging.
     *
     * @param draggable true to enable dragging, false to disable.
     * @return This object for method chaining.
     */
    @NonNull
    public OverlayView<T> setDraggable(boolean draggable) {
        return setDraggable(draggable, new DraggableOnTouchListener<>(this));
    }

    /**
     * Enables or disables moving the view by dragging with a custom listener.
     *
     * @param draggable true to enable dragging, false to disable.
     * @param listener  The custom touch listener for handling drag events.
     * @return This object for method chaining.
     */
    @NonNull
    public OverlayView<T> setDraggable(boolean draggable, DraggableOnTouchListener<T> listener) {
        this.isDraggable = draggable;
        if (draggable) {
            view.setOnTouchListener(listener);
        } else {
            view.setOnTouchListener(null);
        }
        return this;
    }

    /**
     * Sets the width parameter of the view.
     *
     * @param width The width in pixels, or use {@link android.view.ViewGroup.LayoutParams#WRAP_CONTENT}
     *              or {@link android.view.ViewGroup.LayoutParams#MATCH_PARENT}.
     * @return This object for method chaining.
     */
    @NonNull
    public OverlayView<T> setWidth(int width) {
        params.width = width;
        return this;
    }

    /**
     * Sets the height parameter of the view.
     *
     * @param height The height in pixels, or use {@link android.view.ViewGroup.LayoutParams#WRAP_CONTENT}
     *               or {@link android.view.ViewGroup.LayoutParams#MATCH_PARENT}.
     * @return This object for method chaining.
     */
    @NonNull
    public OverlayView<T> setHeight(int height) {
        params.height = height;
        return this;
    }

    /**
     * Sets the placement of the window within the screen as per Gravity.
     *
     * @param gravity The gravity constants from {@link Gravity}.
     * @return This object for method chaining.
     * @see Gravity
     * @see WindowManager.LayoutParams#gravity
     */
    @NonNull
    public OverlayView<T> setGravity(int gravity) {
        params.gravity = GravityCompat.getAbsoluteGravity(gravity,
                ViewCompat.getLayoutDirection(view));
        return this;
    }

    /**
     * Sets the horizontal margin.
     *
     * @param margin The horizontal margin (0 to 1).
     * @return This object for method chaining.
     * @see WindowManager.LayoutParams#horizontalMargin
     */
    @NonNull
    public OverlayView<T> setHorizontalMargin(@FloatRange(from = 0.0, to = 1.0) float margin) {
        params.horizontalMargin = margin;
        return this;
    }

    /**
     * Sets the vertical margin.
     *
     * @param margin The vertical margin (0 to 1).
     * @return This object for method chaining.
     * @see WindowManager.LayoutParams#verticalMargin
     */
    @NonNull
    public OverlayView<T> setVerticalMargin(@FloatRange(from = 0.0, to = 1.0) float margin) {
        params.verticalMargin = margin;
        return this;
    }

    /**
     * Sets the opacity of the view. 0 means completely translucent.
     *
     * @param alpha The opacity value (0 to 1).
     * @return This object for method chaining.
     * @see WindowManager.LayoutParams#alpha
     */
    @NonNull
    public OverlayView<T> setAlpha(@FloatRange(from = 0.0, to = 1.0) float alpha) {
        params.alpha = alpha;
        return this;
    }

    /**
     * Sets the X position of the view.
     *
     * @param x The X position in pixels.
     * @return This object for method chaining.
     * @see WindowManager.LayoutParams#x
     */
    @NonNull
    public OverlayView<T> setX(int x) {
        params.x = x;
        return this;
    }

    /**
     * Gets the X position of the view.
     *
     * @return The X position in pixels.
     * @see WindowManager.LayoutParams#x
     */
    public int getX() {
        return params.x;
    }

    /**
     * Sets the Y position of the view.
     *
     * @param y The Y position in pixels.
     * @return This object for method chaining.
     * @see WindowManager.LayoutParams#y
     */
    @NonNull
    public OverlayView<T> setY(int y) {
        params.y = y;
        return this;
    }

    /**
     * Gets the Y position of the view.
     *
     * @return The Y position in pixels.
     * @see WindowManager.LayoutParams#y
     */
    public int getY() {
        return params.y;
    }

    /**
     * Sets the brightness of the screen.
     *
     * @param screenBrightness The brightness value (0 to 1).
     * @return This object for method chaining.
     * @see WindowManager.LayoutParams#screenBrightness
     */
    @NonNull
    public OverlayView<T> setScreenBrightness(float screenBrightness) {
        params.screenBrightness = screenBrightness;
        return this;
    }

    @VisibleForTesting
    void updateParams() {
        params.type = TYPE_APPLICATION;
        params.flags = FLAG_NOT_FOCUSABLE;

        if (allowViewToExtendOutsideScreen) {
            params.flags |= FLAG_LAYOUT_NO_LIMITS;
            params.flags |= FLAG_LAYOUT_INSET_DECOR;
            params.flags |= FLAG_LAYOUT_IN_SCREEN;
        }

        if (isTouchable || isDraggable) {
            if (allowViewToExtendOutsideScreen && Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
                // On the navigation
                params.type = TYPE_SYSTEM_ERROR;
            } else {
                // Touchable, cannot overlay on the navigation
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    params.type = TYPE_APPLICATION_OVERLAY;
                } else {
                    params.type = TYPE_SYSTEM_ALERT;
                }
            }
        } else {
            params.flags |= FLAG_NOT_TOUCHABLE;
            params.flags |= FLAG_NOT_TOUCH_MODAL;
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                params.type = TYPE_APPLICATION_OVERLAY;
            } else {
                params.type = TYPE_SYSTEM_OVERLAY;
            }
        }

        if (viewScope == ViewScope.ACTIVITY) {
            params.type = TYPE_APPLICATION;
        }
    }

}

File: core/src/main/kotlin/com/nagopy/android/overlayviewmanager/OverlayView.kt
/*
 * Copyright 2026 75py
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.nagopy.android.overlayviewmanager

import android.os.Build
import android.os.Looper
import android.view.View
import android.view.WindowManager
import android.view.Gravity
import androidx.annotation.MainThread
import com.nagopy.android.overlayviewmanager.internal.Logger
import com.nagopy.android.overlayviewmanager.internal.MaximumObscuringOpacity
import com.nagopy.android.overlayviewmanager.internal.OverlayWindowManager
import com.nagopy.android.overlayviewmanager.internal.PassThroughOpacityRegistry

/** A synchronous, main-thread-owned overlay handle. */
public class OverlayView<T : View> internal constructor(
    view: T,
    private val scope: OverlayScope,
    backend: OverlayWindowManager,
    initialSpec: OverlaySpec,
    private val permission: OverlayPermission = OverlayPermission(),
    activityReleaseCallback: (() -> Unit)? = null,
) {
    private var backend: OverlayWindowManager? = backend
    private var ownedView: T? = view
    private var detachListener: View.OnAttachStateChangeListener? = null
    /**
     * Invoked exactly once, from [release], to deregister this handle from the owning Activity's
     * entry in `ActivityOverlayRegistry`. Cleared immediately after invocation so that the closure
     * -- and whatever Activity it captured -- is not retained by a disposed handle.
     */
    private var onActivityRelease: (() -> Unit)? = activityReleaseCallback
    private var effectiveSpec: OverlaySpec = initialSpec
    private var effectiveDragListener: DraggableOnTouchListener<T>? = null
    @Volatile private var currentState: OverlayState = OverlayState.CONFIGURED
    @Volatile private var currentFailure: OverlayFailure? = null
    private var libraryRemovalInProgress: Boolean = false

    init {
        if (initialSpec.touchMode == OverlayTouchMode.DRAGGABLE) {
            effectiveDragListener = DraggableOnTouchListener(this)
        }
        applyTouchListener(effectiveDragListener)
    }

    /** The managed view. It cannot be read once this handle is disposed. */
    @get:MainThread
    public val view: T
        get() {
            requireMainThread()
            return ownedView ?: throw IllegalStateException("OverlayView is disposed.")
        }

    /** The last state established by this handle. This value is safe to read from any thread. */
    public val state: OverlayState
        get() = currentState

    /** The last classified failure. This value is safe to read from any thread. */
    public val lastFailure: OverlayFailure?
        get() = currentFailure

    /** The immutable effective configuration, readable on the main thread even after disposal. */
    @get:MainThread
    public val spec: OverlaySpec
        get() {
            requireMainThread()
            return effectiveSpec
        }

    /** Registers the view synchronously when it is not already attached. */
    @MainThread
    public fun show(): OverlayResult {
        requireMainThread()
        when (currentState) {
            OverlayState.DISPOSED -> return disposedFailure()
            OverlayState.ATTACHED -> return success(changed = false)
            OverlayState.CONFIGURED -> Unit
        }
        val candidate = effectiveSpec
        val managedView = ownedView ?: return failure(OverlayFailure.WINDOW_MANAGER_REJECTED, null)
        if (managedView.parent != null) return failure(OverlayFailure.ALREADY_HAS_PARENT, null)
        if (scope == OverlayScope.APPLICATION && Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && !permission.isGranted(managedView.context)) {
            return failure(OverlayFailure.PERMISSION_DENIED, null)
        }
        admitPassThroughOpacity(candidate, managedView)?.let { return failure(it, null) }
        try {
            backendOrThrow().show(managedView, layoutParams(candidate))
        } catch (exception: Exception) {
            return failure(classify(exception), exception)
        }
        updatePassThroughOpacityRegistration(candidate)
        accept(candidate, nextDragListener(candidate))
        currentState = OverlayState.ATTACHED
        installDetachListener(managedView)
        return success(changed = true)
    }

    /**
     * Applies [spec] synchronously or replaces pending configuration before first show.
     *
     * Unlike [show], this performs no application-scope permission preflight -- neither does
     * [hide] or [dispose]. Only the initial [show] checks [OverlayPermission.isGranted]. If the
     * overlay permission is revoked while [ATTACHED][OverlayState.ATTACHED], the platform reports
     * that on the next window operation: the backend throws [SecurityException], which is
     * classified [OverlayFailure.PERMISSION_DENIED] like any other backend rejection, while the
     * current [state] and the last effective [spec] are preserved unchanged. This mirrors how
     * every other platform rejection here is a later observation rather than a proactive check,
     * and avoids a redundant permission query on every mutation of an already-attached window.
     */
    @MainThread
    public fun update(spec: OverlaySpec): OverlayResult {
        requireMainThread()
        requireApplicationBrightnessIsAbsent(spec)
        if (currentState == OverlayState.DISPOSED) return disposedFailure()
        val nextDragListener = nextDragListener(spec)
        if (currentState == OverlayState.CONFIGURED) {
            if (spec == effectiveSpec) return success(changed = false)
            accept(spec, nextDragListener)
            return success(changed = true)
        }
        if (spec == effectiveSpec) return success(changed = false)
        val managedView = ownedView ?: return failure(OverlayFailure.WINDOW_MANAGER_REJECTED, null)
        admitPassThroughOpacity(spec, managedView)?.let { return failure(it, null) }
        try {
            backendOrThrow().update(managedView, layoutParams(spec))
        } catch (exception: Exception) {
            if (isNotAttached(exception)) {
                reconcileExternalDetach()
                return OverlayResult(currentState, false, OverlayFailure.NOT_ATTACHED, exception)
            }
            return failure(classify(exception), exception)
        }
        updatePassThroughOpacityRegistration(spec)
        accept(spec, nextDragListener)
        return success(changed = true)
    }

    /** Removes the view synchronously. A not-attached result is a successful reconciliation. */
    @MainThread
    public fun hide(): OverlayResult {
        requireMainThread()
        if (currentState == OverlayState.DISPOSED) return success(changed = false, preserveDiagnostic = true)
        if (currentState == OverlayState.CONFIGURED) return success(changed = false, preserveDiagnostic = true)
        val managedView = ownedView ?: return failure(OverlayFailure.WINDOW_MANAGER_REJECTED, null)
        libraryRemovalInProgress = true
        try {
            backendOrThrow().hide(managedView)
        } catch (exception: Exception) {
            libraryRemovalInProgress = false
            if (isNotAttached(exception)) {
                reconcileExternalDetach()
                return successfulReconciliation()
            }
            return failure(classify(exception), exception)
        }
        libraryRemovalInProgress = false
        currentState = OverlayState.CONFIGURED
        removeDetachListener(managedView)
        PassThroughOpacityRegistry.unregister(this)
        return success(changed = true)
    }

    /** Releases owned resources after a successful synchronous removal. */
    @MainThread
    public fun dispose(): OverlayResult {
        requireMainThread()
        if (currentState == OverlayState.DISPOSED) return success(changed = false, preserveDiagnostic = true)
        if (currentState == OverlayState.ATTACHED) {
            val managedView = ownedView ?: return failure(OverlayFailure.WINDOW_MANAGER_REJECTED, null)
            libraryRemovalInProgress = true
            try {
                backendOrThrow().hide(managedView)
            } catch (exception: Exception) {
                libraryRemovalInProgress = false
                if (isNotAttached(exception)) {
                    reconcileExternalDetach()
                    release(managedView)
                    return successfulReconciliation(OverlayState.DISPOSED)
                }
                return failure(classify(exception), exception)
            }
            libraryRemovalInProgress = false
            release(managedView)
            return success(changed = true)
        }
        ownedView?.let(::release)
        return success(changed = true, preserveDiagnostic = true)
    }

    /**
     * Forced cleanup for the owning Activity's `onActivityDestroyed`. If [ATTACHED][OverlayState.ATTACHED],
     * attempts exactly one `removeViewImmediate`; regardless of that outcome, releases every owned
     * reference and transitions to [DISPOSED][OverlayState.DISPOSED]. A failed removal is classified
     * with the existing [classify] mapping and retained as [lastFailure] -- the exception itself is
     * never retained. `DISPOSED` here means library ownership is released, not proof that the
     * framework removal succeeded. Idempotent: a no-op once already [DISPOSED][OverlayState.DISPOSED].
     */
    @MainThread
    internal fun disposeForActivityDestruction() {
        requireMainThread()
        if (currentState == OverlayState.DISPOSED) return
        val managedView = ownedView
        if (currentState == OverlayState.ATTACHED && managedView != null) {
            libraryRemovalInProgress = true
            try {
                backendOrThrow().hide(managedView)
            } catch (exception: Exception) {
                currentFailure = classify(exception)
                Logger.w(exception, "Forced Activity-destruction removeViewImmediate failed; classified as %s", currentFailure)
            } finally {
                libraryRemovalInProgress = false
            }
        }
        managedView?.let(::release)
    }

    private fun installDetachListener(managedView: T) {
        val listener = object : View.OnAttachStateChangeListener {
            override fun onViewAttachedToWindow(v: View) = Unit
            override fun onViewDetachedFromWindow(v: View) {
                if (!libraryRemovalInProgress && currentState == OverlayState.ATTACHED) reconcileExternalDetach()
            }
        }
        detachListener = listener
        managedView.addOnAttachStateChangeListener(listener)
    }

    private fun removeDetachListener(managedView: T) {
        detachListener?.let(managedView::removeOnAttachStateChangeListener)
        detachListener = null
    }

    private fun reconcileExternalDetach() {
        currentState = OverlayState.CONFIGURED
        currentFailure = OverlayFailure.NOT_ATTACHED
        ownedView?.let(::removeDetachListener)
        PassThroughOpacityRegistry.unregister(this)
    }

    private fun release(managedView: T) {
        removeDetachListener(managedView)
        managedView.setOnTouchListener(null)
        effectiveDragListener = null
        ownedView = null
        backend = null
        currentState = OverlayState.DISPOSED
        PassThroughOpacityRegistry.unregister(this)
        val activityRelease = onActivityRelease
        onActivityRelease = null
        activityRelease?.invoke()
    }

    private fun accept(spec: OverlaySpec, dragListener: DraggableOnTouchListener<T>?) {
        effectiveSpec = spec
        effectiveDragListener = dragListener
        applyTouchListener(dragListener)
    }

    private fun nextDragListener(spec: OverlaySpec): DraggableOnTouchListener<T>? {
        if (spec.touchMode != OverlayTouchMode.DRAGGABLE) return null
        if (effectiveSpec.touchMode == OverlayTouchMode.DRAGGABLE && effectiveDragListener != null) {
            return effectiveDragListener
        }
        return DraggableOnTouchListener(this)
    }

    private fun applyTouchListener(listener: DraggableOnTouchListener<T>?) {
        ownedView?.setOnTouchListener(listener)
    }

    private fun layoutParams(value: OverlaySpec): WindowManager.LayoutParams = WindowManager.LayoutParams().apply {
        width = value.width
        height = value.height
        gravity = Gravity.getAbsoluteGravity(value.gravity, view.layoutDirection)
        x = value.x
        y = value.y
        horizontalMargin = value.horizontalMargin
        verticalMargin = value.verticalMargin
        alpha = value.alpha
        format = android.graphics.PixelFormat.TRANSLUCENT
        flags = WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE
        if (value.allowOutsideBounds) flags = flags or WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS or WindowManager.LayoutParams.FLAG_LAYOUT_INSET_DECOR or WindowManager.LayoutParams.FLAG_LAYOUT_IN_SCREEN
        if (value.touchMode == OverlayTouchMode.PASS_THROUGH) flags = flags or WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE
        type = if (scope == OverlayScope.ACTIVITY) WindowManager.LayoutParams.TYPE_APPLICATION else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY else WindowManager.LayoutParams.TYPE_SYSTEM_ALERT
        if (scope == OverlayScope.ACTIVITY && value.screenBrightness != null) screenBrightness = value.screenBrightness
    }

    private fun success(changed: Boolean, preserveDiagnostic: Boolean = false): OverlayResult {
        if (!preserveDiagnostic) currentFailure = null
        return OverlayResult(currentState, changed, null, null)
    }

    private fun successfulReconciliation(finalState: OverlayState = OverlayState.CONFIGURED): OverlayResult {
        currentState = finalState
        currentFailure = OverlayFailure.NOT_ATTACHED
        return OverlayResult(finalState, false, null, null)
    }

    private fun failure(kind: OverlayFailure, cause: Throwable?): OverlayResult {
        currentFailure = kind
        return OverlayResult(currentState, false, kind, cause)
    }

    private fun disposedFailure(): OverlayResult = OverlayResult(
        currentState,
        false,
        OverlayFailure.WINDOW_MANAGER_REJECTED,
        null,
    )

    private fun backendOrThrow(): OverlayWindowManager = checkNotNull(backend) { "OverlayView is disposed." }

    /**
     * The Android 12+ (API 31+) opacity-budget admission check. Applicable only to an
     * application-scope [OverlayTouchMode.PASS_THROUGH] candidate that opts into
     * [CrossUidPassThrough.WHEN_SYSTEM_ALLOWS]. Runs on every [show]/[update] that reaches this
     * point -- callers already skip a true no-op spec before calling here -- including an x/y,
     * size, gravity, or margin-only change: the platform maximum can change between operations, so
     * it is queried fresh every time rather than cached or skipped for a non-alpha change. Returns
     * the failure to report, or `null` to proceed with the platform call.
     */
    private fun admitPassThroughOpacity(candidate: OverlaySpec, managedView: T): OverlayFailure? {
        if (scope != OverlayScope.APPLICATION) return null
        if (candidate.touchMode != OverlayTouchMode.PASS_THROUGH) return null
        if (candidate.crossUidPassThrough != CrossUidPassThrough.WHEN_SYSTEM_ALLOWS) return null
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S) return null
        val maximum = try {
            MaximumObscuringOpacity.get().maximumObscuringOpacityForTouch(managedView.context)
        } catch (exception: Exception) {
            null
        } ?: return OverlayFailure.PASS_THROUGH_UNSUPPORTED
        val combined = PassThroughOpacityRegistry.combinedOpacityExcluding(this, candidate.alpha)
        return if (combined > maximum) OverlayFailure.PASS_THROUGH_OPACITY_EXCEEDED else null
    }

    /** Registers or drops this handle's alpha in [PassThroughOpacityRegistry] after a successful backend call. */
    private fun updatePassThroughOpacityRegistration(spec: OverlaySpec) {
        if (scope == OverlayScope.APPLICATION && spec.touchMode == OverlayTouchMode.PASS_THROUGH) {
            PassThroughOpacityRegistry.register(this, spec.alpha)
        } else {
            PassThroughOpacityRegistry.unregister(this)
        }
    }

    private fun requireApplicationBrightnessIsAbsent(spec: OverlaySpec) {
        require(scope != OverlayScope.APPLICATION || spec.screenBrightness == null) {
            "screenBrightness is supported only for activity overlays."
        }
    }

    private fun classify(exception: Exception): OverlayFailure = when {
        isNotAttached(exception) -> OverlayFailure.NOT_ATTACHED
        exception is WindowManager.BadTokenException -> OverlayFailure.INVALID_WINDOW_TOKEN
        exception is SecurityException -> OverlayFailure.PERMISSION_DENIED
        else -> OverlayFailure.WINDOW_MANAGER_REJECTED
    }

    private fun isNotAttached(exception: Exception): Boolean = exception is IllegalArgumentException && exception.message?.contains("not attached", ignoreCase = true) == true

    private fun requireMainThread() {
        val main = Looper.getMainLooper()
        check(main != null && Looper.myLooper() === main) { "OverlayView must be used on the main thread." }
    }

}
